### ISSUE TITLE
[Bug]: Discrete Graphics option seems to do the opposite of what it says

### ISSUE THREAD

**Issue by michaelneverwins** [bottlesdevs/Bottles#2967]
### Describe the bug

Today I created a new bottle to install and run the GOG version of _Blood II_. My PC happens to have an Intel CPU with integrated graphics as well as a discrete AMD GPU, and I noticed while checking the game's display settings that the two displays listed were DirectDraw HAL and Intel HD Graphics. This is not what I expected, because the "Discrete Graphics" option was enabled in the bottle settings. When I disabled that and started _Blood II_ again, the game's display settings showed an AMD Radeon HD option instead of Intel HD Graphics.

![blood-discrete-on](https://github.com/bottlesdevs/Bottles/assets/73746653/24d7743d-102a-4db0-b772-a2cdb7f6e140)
![blood-discrete-off](https://github.com/bottlesdevs/Bottles/assets/73746653/d449e219-91bc-4723-9386-d55dc4f9ffd6)

To determine whether the display names in the game's settings were accurate, I tried running the game both ways with a temperature monitor running. Sure enough, I got higher CPU temperatures when the game indicated Intel HD Graphics (i.e. when Bottles' "Discrete Graphics" setting was enabled) and slightly higher GPU temperatures when the game indicated AMD Radeon HD (i.e. when Bottles' "Discrete Graphics" setting was disabled). So I don't think the game's menu is just confused about what each display is called.

In other words, it seems that the game detects and uses my integrated graphics chip when the bottle's "Discrete Graphics" option is enabled, and detects my discrete graphics card when the bottle's "Discrete Graphics" option is disabled. This is the opposite of what I assume it should do.

I don't know of any other games which explicitly indicate which graphics card they are using, and thus I don't know if this bug occurs with every game. I suspect that it does not. The few other games I've set up in Bottles — the itch.io version of _Halloween Forever_, the freeware game _Lyle in Cube Sector_, and a DVD copy of _Clive Barker's Jericho_ — fail to run at all with the "Discrete Graphics" option switched off. It seems more likely that they would fail to run with integrated graphics than with my AMD card, so for those games, I suspect that my discrete AMD card actually is being used with the "Discrete Graphics" option turned on. For _Clive Barker's Jericho_ in particular, I don't think I would get the performance I'm seeing if the bottle were erroneously using my integrated graphics instead.

### To Reproduce

1. Create a new "Gaming" bottle `Blood2`.
2. Download the GOG version of _Blood II_ (`setup_blood2_2.0.0.5.exe`).
3. Copy `setup_blood2_2.0.0.5.exe` into `~/.var/app/com.usebottles.bottles/data/bottles/bottles/Blood2/drive_c/users/$USER/Downloads/`.
4. Click "Run Executable" and select `~/.var/app/com.usebottles.bottles/data/bottles/bottles/Blood2/drive_c/users/$USER/Downloads/setup_blood2_2.0.0.5.exe` to run the _Blood II_ installer.
5. Click "Add Shortcuts..." and select `~/.var/app/com.usebottles.bottles/data/bottles/bottles/Blood2/drive_c/GOG Games/Blood II - The Chosen/BLOOD2.EXE`.
6. Click the `BLOOD2` shortcut.
7. In the launcher window, click "Display..." and observe the listed displays.
8. Close the game.
9. Toggle the bottle's "Discrete Graphics" option under "Settings".
10. Repeat steps 6 and 7.

### Package

Flatpak from Flathub

### Distribution

Linux Mint 21.1

### Debugging Information

```shell
Official Package: true
Version: '51.6'
DE/WM: cinnamon
Display:
    X.org: true
    X.org (port): :99.0
    Wayland: false
Graphics:
    vendors:
        amd: &amp;id002
            vendor: amd
            envs:
                DRI_PRIME: '1'
            icd: /usr/lib/x86_64-linux-gnu/GL/vulkan/icd.d/radeon_icd.x86_64.json:/usr/lib/i386-linux-gnu/GL/vulkan/icd.d/radeon_icd.i686.json
        intel: &amp;id001
            vendor: intel
            envs:
                DRI_PRIME: '1'
            icd: /usr/lib/x86_64-linux-gnu/GL/vulkan/icd.d/intel_hasvk_icd.x86_64.json:/usr/lib/x86_64-linux-gnu/GL/vulkan/icd.d/intel_icd.x86_64.json:/usr/lib/i386-linux-gnu/GL/vulkan/icd.d/intel_hasvk_icd.i686.json:/usr/lib/i386-linux-gnu/GL/vulkan/icd.d/intel_icd.i686.json
    prime:
        integrated: *id001
        discrete: *id002
Kernel:
    Type: Linux
    Version: 5.15.0-76-generic
Disk:
    Total: 4103446528
    Free: 4103282688
RAM:
    MemTotal: 7.6GiB
    MemAvailable: 5.8GiB
Bottles_envs: null
```

**Comment by Perkolator** [comment]
I have seen settings for a bottle getting mixed up. The GUI shows something but in reality the setting for the bottle is the other way. Sadly there seems to be bugs like this in Bottles.

I haven't used Bottles for long but from the get go, while trying different settings and trying to learn, I have been scratching my head quite many times. For example, I have this one game (Stasis - Bone Totem) which worked fine except the "cinematics" wouldn't play at all. So I tried different settings and runners and nothing seemed to make the cinematics work (or even not start the game at all), so I decided to put the settings back to what they were, and suddenly the game wouldn't run anymore! Exactly the same settings. I have seen this "changing bottle settings messes up the bottle environment" happen many times (it's really hard to start learning to use the program and all that it encompasses when it's chaos with the settings, let alone trying to create re-produce steps for an issue).

I have one example of a GUI setting showing the setting wrong. I haven't posted an issue yet, or checked if it has already one posted. I have a big list of bugs and stuff to post here, if they aren't already posted, too busy with something else (though still occasionally brought back here).

In short; in my experience, the less you touch the bottle settings, the better. And NEVER use a bottle for multiple programs/games, always just one program/game per bottle (if you need to set different settings/install dependencies for a new program/game, you'll just probably mess up the whole bottle and the already working program/game).

**Comment by orowith2os** [comment]
@Perkolator there's no need for that kind of comment. File a bug and move on, and maybe consider that this isn't an issue specific to Bottles. Wine is inherently hacky. 

As for the topic at hand: it *looks* like the internals are working fine, I can't think of a reason as to why this is happening. Bottles shouldn't be cancelling out DRI_PRIME like this. I will need to investigate this some more.

**Comment by Perkolator** [comment]
> there's no need for that kind of comment

Hmm, it was just my experience and I thought it was related to this issue. I don't see anything rude about it. Maybe it was too blunt? I guess that's the culture of where I'm from, we don't sugarcoat, we say it directly. Didn't mean to offend anyone. If I wouldn't be interested about the fate of Bottles, I wouldn't be active here.

> File a bug

Like I wrote, it's really really hard when things happen randomly. I have tried to catch these but no luck yet. I'm not new to debugging and I've been scratching my head with these problems. Though, I have one bug/problem, that I can re-produce but haven't reported yet, which backs up my suggestion to never use a bottle for more than one program.

**Comment by michaelneverwins** [comment]
A bit more information about my system:

I just noticed that running `glxinfo | grep "OpenGL renderer"` prints
```
OpenGL renderer string: AMD Radeon HD 7900 Series (tahiti, LLVM 15.0.7, DRM 3.42, 5.15.0-76-generic)
```
whereas running `DRI_PRIME=1 glxinfo | grep "OpenGL renderer"` prints
```
OpenGL renderer string: Mesa Intel(R) HD Graphics 4600 (HSW GT2)
```
which surprised me a bit. I was under the impression that the environment variable `DRI_PRIME=1` was for selecting the discrete GPU, but I don't actually know much about this subject. Maybe it's technically for selecting the non-default graphics but it happens to be the case that people don't use it unless their discrete GPU is not the default...?

Anyway, I don't know if that's related to the reported bug, but it seems like it might be.

If nobody else is able to reproduce the Bottles bug, then maybe it's just a "me" problem, but I don't think I did anything really weird to my system except for [using kernel parameters `radeon.si_support=0 amdgpu.si_support=1` to enable `amdgpu` support for my GPU](https://wiki.archlinux.org/title/AMDGPU#Set_module_parameters_in_kernel_command_line).

**Comment by orowith2os** [comment]
@michaelneverwins it could be that your AMD GPU doesn't support the necessary features for DXVK 2.x. There *is* a reason why amdgpu isn't enabled by default on it, after all. Maybe try reproducing with the latest 1.x DXVK version, and see if this still happens.

**Comment by michaelneverwins** [comment]
I don't expect this old GPU to have perfect Vulkan support, etc., but I didn't think missing DXVK features would be responsible for Bottles giving the wrong display to a game. In any case, is there an easy way to try DXVK 1.x in Bottles? The only options I see in the bottle settings now are `dxvk-2.1-1-977669c` and `Disabled`.

For what it's worth, if I disable DXVK in the bottle settings, I still get the same result described in the original post: with "Discrete Graphics" enabled, the game's display settings show my Intel HD graphics, and with "Discrete Graphics" disabled, the game's display settings show my AMD Radeon HD graphics.

(Simply turning off "Discrete Graphics" if I want to use my AMD card is the obvious workaround, by the way. I submitted this bug report because I figured it would be of interest to the Bottles developers, but obviously nothing is preventing me from playing the game with discrete graphics; I just have to remember that the "Discrete Graphics" setting is flip-flopped for some reason.)

**Comment by Kinsteen** [comment]
You can go in Bottles' settings and download whichever component you'd like

**Comment by michaelneverwins** [comment]
>You can go in Bottles' settings and download whichever component you'd like

Oh, thanks. I see it. I should have tried a bit harder to find it before asking.

**Comment by michaelneverwins** [comment]
With the bottle set to use `dxvk-1.10.3`, I still get the same result with regard to bottle's the "Discrete Graphics" switch and the game's display menu.

**Comment by orowith2os** [comment]
> I don't expect this old GPU to have perfect Vulkan support, etc., but I didn't think missing DXVK features would be responsible for Bottles giving the wrong display to a game.

It would be, as it would then fallback to, and run on, a GPU which supports it - your Intel, most likely. 

Another thing to try would be disabling amdgpu, and seeing what happens then. Vulkan support would most likely get dropped from the AMD card, but then you can see if you can reproduce it with WineD3D too, and be sure it's not a Vulkan thing.

**Comment by michaelneverwins** [comment]
Question: Does the "Discrete Graphics" option actually make the bottle prefer the non-default GPU?

Since upgrading to Linux Mint 21.2 today, I'm noticing that my discrete graphics card being my system's default — as demonstrated in my earlier comment https://github.com/bottlesdevs/Bottles/issues/2967#issuecomment-1646945947 — is apparently problematic in general.

To use an example unrelated to Bottles, the desktop entry for Super Tux Kart contains `PrefersNonDefaultGPU=true`, and when I launch the game using the menu shortcut defined by that file, performance is bad and temperature monitoring very clearly shows it's using my integrated graphics. If I change it to `PrefersNonDefaultGPU=false`, then performance is good and temperature monitoring suggests the game is actually using my discrete graphics card. (I didn't have any problems with this game on Linux Mint 21.1, for whatever reason, maybe because the use of the `PrefersNonDefaultGPU` flag wasn't implemented in that version.) It seems the Linux community has decided that the "non-default" GPU is for gaming, which implies that my system's apparent use of discrete graphics by default is flat-out wrong.

I also had another possibly related problem launching Steam after upgrading to Mint 21.2, which was fixed by setting `PrefersNonDefaultGPU=false` in its desktop entry as explained [in this topic on the Linux Mint forums](https://forums.linuxmint.com/viewtopic.php?t=400351).)

Maybe I should be figuring out how to make my integrated graphics the system default (and if I do then I'll re-test _Blood 2_ with Bottles and see what happens), but then I'm not sure what will happen to running games that _aren't_ using a `PrefersNonDefaultGPU=true` flag or similar.

**Comment by Kinsteen** [comment]
I think that you have a weird system that makes DRI_PRIME work in weird ways, with GPUs mixed up somehow... It's the first time I've seen this type of behavior, but it's definitely not Bottles fault as we use the DRI_PRIME variable, like you used it in your glxinfo test

**Comment by michaelneverwins** [comment]
If Bottles' "Discrete Graphics" switch is just setting `DRI_PRIME=1`, then maybe what I'm seeing with _Blood 2_ makes sense in light of my `glxinfo` output. If it really is that simple, then I guess this can be closed.

Although, in that case, I still don't understand what I'm seeing with other games, such as _Clive Barker's Jericho_:

* DXVK set to `dxvk-2.1-1-977669c`, "Discrete Graphics" disabled: _Jericho_ won't launch at all.
* DXVK set to `dxvk-2.1-1-977669c`, "Discrete Graphics" enabled: _Jericho_ has good performance, and temperature monitoring suggests it is using my discrete GPU.
* DXVK set to `dxvk-1.10.3`, "Discrete Graphics" disabled: _Jericho_ has poor performance, and my discrete GPU temperature stays low while CPU temperature skyrockets.
* DXVK set to `dxvk-1.10.3`, "Discrete Graphics" enabled: _Jericho_ performance is good again, and the GPU temperature goes up.

This seems to show the "Discrete Graphics" setting actually doing what it says it does (with perhaps my integrated graphics simply not supporting DXVK 2.1). I wouldn't expect to to see this, if the explanation were simply that my discrete GPU is the default and `DRI_PRIME` thinks my integrated graphics chip is the discrete one.

Scrolling back up to the system information I included with original post, I see
```
            envs:
                DRI_PRIME: '1'
```
appearing both under `vendor: amd` and under `vendor: intel`, but I don't know if that's significant.

**Comment by orowith2os** [comment]
You can chalk a lot of that up to Vulkan stuff; older GPUs are weird. 

> Question: Does the "Discrete Graphics" option actually make the bottle just prefer the non-default GPU?

I believe so, yes. It just sets DRI_PRIME=1. 0 means the default, 1 means non-default; this gets messy when you see more than one GPU, but w/e. 

> appearing both under vendor: amd and under vendor: intel, but I don't know if that's significant.

Now *that* might be. I'm not sure if this matters much if you can't reproduce with other titles though. Also try plain Wine and other methods to see if the DRI_PRIME issue shows up too.

---

On another topic... I misread your post in https://github.com/bottlesdevs/Bottles/issues/2967#issuecomment-1646945947. I'm sorry. Yeah, I think your system might be messed up somewhere; the issue is we don't know who's causing it.

**Comment by michaelneverwins** [comment]
It occurs to me that _Blood II_ uses Direct3D 6 (at least according to [PC Gaming Wiki](https://www.pcgamingwiki.com/wiki/Blood_II:_The_Chosen#API)) while _Clive Barker's Jericho_ uses Direct3D 9 ([same source](https://www.pcgamingwiki.com/wiki/Clive_Barker%27s_Jericho#API)). DXVK is for Direct3D 9/10/11 only. Therefore _Blood II_ would, I assume, be running under WineD3D, while _Jericho_ would actually be running under DXVK. Perhaps the apparent inconsistency in the "Discrete Graphics" option's effect (or in the handling of `DRI_PRIME=1`) is actually a difference between those two Direct3D translation layers. It could be that WineD3D (or OpenGL) handles `DRI_PRIME=1` by simply selecting the non-default GPU, while DXVK (or Vulkan) handles `DRI_PRIME=1` by selecting the discrete one even if it's the default. Or maybe they try to do the same thing but, on my apparently unusual configuration, disagree about which is non-default or which is discrete.

To test the hypothesis that it's an OpenGL-versus-Vulkan discrepancy, I tried something with GZDoom, because it's native (read: easy to run) and lets me choose the renderer. In OpenGL mode, GZDoom runs well by default, but then running it with `DRI_PRIME=1` degrades performance (most noticeably in the screen-melt effect which slows down) and also increases CPU temperatures (which I've taken as the indication that it's using my integrated graphics). In Vulkan mode, GZDoom seems to have the same performance regardless of whether I run it with `DRI_PRIME_1`, so I assume it's using my discrete GPU in both cases. GZDoom's console output confirms it:
* Without `DRI_PRIME=1`, in OpenGL mode: `GL_RENDERER: AMD Radeon HD 7900 Series (tahiti, LLVM 15.0.7, DRM 3.42, 5.15.0-76-generic)`
* With `DRI_PRIME=1`, in OpenGL mode: `GL_RENDERER: Mesa Intel(R) HD Graphics 4600 (HSW GT2)`
* Without `DRI_PRIME=1`, in Vulkan mode: `Vulkan device: AMD Radeon HD 7900 Series (RADV TAHITI)`
* With `DRI_PRIME=1`, in Vulkan mode: `Vulkan device: AMD Radeon HD 7900 Series (RADV TAHITI)`

This isn't quite what I was seeing in Bottles running Windows games, because _Jericho_ actually did need the "Discrete Graphics" option enabled in order to run well (perhaps because it was using DXVK instead of Vulkan directly) whereas GZDoom's Vulkan mode runs the same with and without `DRI_PRIME=1`. However, this shows there's a difference between OpenGL and Vulkan. I mentioned _Super Tux Kart_ earlier; it also uses OpenGL, and when I run it from the terminal, the logging confirms that it also selects my integrated graphics when executed with `DRI_PRIME=1`. (I don't think it has a Vulkan mode I can test.)

Anyway, sorry for going so far down the rabbit hole here, but this testing has convinced me that this is not a Bottles bug. Rather, I think it's just an issue with the way OpenGL handles `DRI_PRIME` on my system (or perhaps any system on which the discrete GPU is the default one). So I'm going to close this issue.

By the way, as for the recommendation from @orowith2os:
>Another thing to try would be disabling amdgpu, and seeing what happens then. Vulkan support would most likely get dropped from the AMD card, but then you can see if you can reproduce it with WineD3D too, and be sure it's not a Vulkan thing.

I disabled `amdgpu` and then tested Bottles with _Blood 2_ again, and I got the same result as with `amdgpu` enabled. I guess that's no surprise if we know it was using WineD3D all along anyway. It's definitely not a Vulkan thing.

**Comment by michaelneverwins** [comment]
Although this is now closed as "not a Bottles bug", I just want to post some more details here for the benefit of anyone else who sees the same behavior in Bottles — or, more generally, anyone who has the same issue with `DRI_PRIME`/`PrefersNonDefaultGPU` as a result of the discrete GPU being the default on a system with both discrete and integrated graphics. This issue might come up in a web search, and finding relevant discussion of a problem with no resolution at the end is always frustrating.

A couple of days ago, I described what I was seeing with `DRI_PRIME` and `PrefersNonDefaultGPU` on ValveSoftware/steam-for-linux#9383, because the Steam bug discussed in that issue is somehow related to `PrefersNonDefaultGPU` and I thought that problem might be related to my flip-flopped `DRI_PRIME` behavior. I no longer believe that's the case, but there were some generally helpful comments about `DRI_PRIME` ([here](https://github.com/ValveSoftware/steam-for-linux/issues/9383#issuecomment-1647221383), [here](https://github.com/ValveSoftware/steam-for-linux/issues/9383#issuecomment-1649200329), and [here](https://github.com/ValveSoftware/steam-for-linux/issues/9383#issuecomment-1649398686)) in response to my somewhat off-topic questions.

I don't know how true it is that `DRI_PRIME` is primarily for laptops, as described there, but my system is actually a desktop PC which just happens to have both a dedicated graphics card as well as a CPU with integrated graphics. For as long as I've used this PC, my monitors have been plugged directly into the dedicated video card (as opposed to using the HDMI, DVI, etc. ports that go to the motherboard instead). I never gave it a second thought, and so it took me quite a while to realize that perhaps this was the reason that Linux Mint used the discrete graphics as default. It was.

Simply plugging my monitors into the motherboard's ports was not sufficient, however, because then I had no desktop graphics at all. I had to go to my machine's BIOS settings and change "Initial Display Output" from "PCIe 1 Slot" to "IGFX". (I'm sure the nomenclature here will vary from one system to another.) Then booting with my monitors plugged into the motherboard worked. I have my PC in this configuration now, and Linux Mint is now reporting my integrated Intel HD graphics as the default, with my discrete AMD Radeon HD graphics as the non-default.

In other words, running `glxinfo | grep "OpenGL renderer"` now prints
```
OpenGL renderer string: Mesa Intel(R) HD Graphics 4600 (HSW GT2)
```
while running `DRI_PRIME=1 glxinfo | grep "OpenGL renderer"` prints
```
OpenGL renderer string: AMD Radeon HD 7900 Series (tahiti, LLVM 15.0.7, DRM 3.42, 5.15.0-76-generic)
```
which is the behavior I think most people expect from `DRI_PRIME` based on what I've read.

Additionally, all of the related oddities are also cleared up.
* GZDoom in OpenGL mode now uses integrated graphics by default, and discrete graphics when executed with `DRI_PRIME=1`.
* GZDoom in Vulkan mode still uses my discrete graphics regardless of `DRI_PRIME`.
* Super Tux Kart, launched from its menu shortcut containing `PrefersNonDefaultGPU=true`, now uses discrete graphics. (If I change it to `PrefersNonDefaultGPU=false`, it uses integrated graphics.)
* When right-clicking a game in Linux Mint's application menu, the "Run with dedicated graphics" option actually seems to do what it says. (I didn't mention this option before, but it had previously been doing the opposite of what it says, at least for OpenGL games; it probably just sets `DRI_PRIME=1`.)

Finally, I'm now seeing the expected behavior in Bottles with _Blood II_. With the bottle's "Discrete Graphics" option enabled, the game's display settings show my AMD Radeon HD card; with "Discrete Graphics" disabled, the game's display settings show my Intel HD Graphics chip. I haven't taken the time to ensure that _Jericho_ still behaves as expected under DXVK, but I would bet that it does.

So the mystery is solved. My discrete graphics card was the default because my monitors were plugged into it, and anything setting `DRI_PRIME=1` will apparently tell OpenGL to use the non-default graphics card.

I'm still not entirely happy with the way that the non-default graphics card is ubiquitously assumed to be the discrete one. It turns out that the apparently problematic aspect of my system configuration (i.e. monitors plugged into the GPU) really isn't that weird, so surely there are others with non-default integrated graphics as well. I would say that the "Discrete Graphics" option in Bottles should be called "Non-Default Graphics" or similar, but this wording would probably just be more confusing for the majority of users, and doesn't really get at the root of the problem. Even in the case of a similar option being named in an arguably more accurate way (i.e. the `PrefersNonDefaultGPU` line in desktop entries), the usage still implies that "non-default" _means_ "discrete" (i.e. gaming-related shortcuts using `PrefersNonDefaultGPU=true` imply that the non-default GPU is for gaming).

In any case, now that I understand that my system isn't _broken_ in some way, I might plug my monitors back into the GPU and just deal with some programs assuming the wrong default graphics. Or maybe I'll keep my integrated graphics as the default for a while and see how it goes.
